src/org/python/core/JavaProxyList.java

```diff
@@ -42,7 +42,39 @@ protected List newList() {
                 throw Py.JavaError(e);
             }
         }
-    }
+
+        protected boolean isEqual(PyObject other) {
+            List jList = asList();
+            if (other.getType().isSubType(PyList.TYPE)) {
+                PyList oList = (PyList) other;
+                if (jList.size() != oList.size()) {
+                    return false;
+                }
+                for (int i = 0; i < jList.size(); i++) {
+                    if (!Py.java2py(jList.get(i))._eq(oList.pyget(i)).__nonzero__()) {
+                        return false;
+                    }
+                }
+                return true;
+            } else {
+                Object oj = other.getJavaProxy();
+                if (oj instanceof List) {
+                    List oList = (List) oj;
+                    if (jList.size() != oList.size()) {
+                        return false;
+                    }
+                    for (int i = 0; i < jList.size(); i++) {
+                        if (!Py.java2py(jList.get(i))._eq(Py.java2py(oList.get(i))).__nonzero__()) {
+                            return false;
+                        }
+                    }
+                    return true;
+                } else {
+                    return false;
```

Comment:
It may be right to refactor the code like this, but it has to be in a way that retains the possibility of returning `null, meaning "don't know".

Reply:
This is copying the style already used in JavaProxySet and I think its a little nicer. I'm not clear why we need to say Don't know to reach the false at the end we are being asked if we are equal to an object that is neither a Python list or a Java list therefore we know we can't be equal? Maybe i'm missing something? Could you provide an example where the previous code would behave differently to this in this respect?

Reply:
If we return don't know, then the other object type will be asked, which is allowed to have its own idea of equality means, even equality between its instances and those of PyList (e.g. it does case insensitve comparisons or something). If this ither type were a (Python) sub-type of PyList, it would be asked first, and a "don't know" return here would end up as a False, after one last attempt using `__cmp__`. However, even when it is not a sub-type, it has to be given a go at answering. The logic is here:
https://github.com/jythontools/jython/blob/bc9d6ea8497b39fc52ed358bf4f844197ee48cef/src/org/python/core/PyObject.java#L1464-L1472